src/redux-middleware/pullBeforeMove.ts

```diff
@@ -0,0 +1,35 @@
+import { ThunkMiddleware } from 'redux-thunk'
+import pull from '../action-creators/pull'
+import { isPending, pathExists, simplifyPath } from '../selectors'
+import { hashContext, head, parentOf, pathToContext } from '../util'
+import { State } from '../util/initialState'
+
+/**
+ * Middleware that fetches any pending descendants before a thought-move operation.
+ */
+const pullBeforeMove: ThunkMiddleware<State> = ({ getState, dispatch }) => {
+
+  return next => async action => {
+
+    if (action.type === 'existingThoughtMove') {
+      const state = getState()
+
+      const oldContext = pathToContext(simplifyPath(state, action.oldPath))
+
+      const newContext = pathToContext(simplifyPath(state, action.newPath))
+
+      const updatedNewContext = [...parentOf(newContext), head(oldContext)]
+      const conflictedPath = pathExists(state, updatedNewContext)
+
+      const isNewContextPending = conflictedPath && isPending(state, newContext)
+      if (isNewContextPending) {
+        const toPull = { [hashContext(newContext)]: newContext }
+        await dispatch(pull(toPull, { maxDepth: Infinity }))
```

Comment:
We'll want to limit the depth of the pull, otherwise there could be a large number of thoughts pulled unnecessarily.

Reply:
In order to optimize this and pull only until a certain depth, we need to know the full conflicting path, which cannot be determined without loading the source context's tree, which is [loaded after the reducer](https://github.com/cybersemics/em/blob/a44a411697efb8112440386c3f4cb7f343bca746/src/reducers/existingThoughtMove.ts#L359) `existingThoughtMove` gets triggered.
For instance, 
`
 - a
    - b
      - c
        - d1
        - d2
          - e1 (pending)
          - e2 (pending)
 - p
    - b
      - c
        - d2
          - e3 (pending)
          - e4 (pending 
`

At the time of running the middleware, we'd only know about the conflict `['p', 'b', 'c']`, and any conflicts beyond that level cannot be predicted, unless we load the source path in the middleware. 
Let me know what you think.

Reply:
Okay, that makes sense. If there is a deep conflict we need to pull source and destination.

There is a case where the source is very shallow (e.g. 2 levels) and the destination is very deep (e.g. 10 levels). In that case, it would be best to set `maxDepth: 2` since we know there can be no conflicts beyond that.

This may sound like an edge case, but it's actually a pretty common use case in my experience. Let's say I'm moving a few thoughts to a different part of my thoughtspace. I will probably have to move them up a few levels and even into the root before moving them back down the hierarchy to where I wanted them (although Bijay just finished up expand-on-drag which makes it possible to do in a single move). Pulling the root with `maxDepth: Infinity` will crash the app.

We may have to pull the source, count the depth, then pull the destination up to that depth.